### PR TITLE
Added Support for specifying range on random number in properties

### LINKED ISSUE #5291: [Feature] support "configurable" random in PropertyResolver / or allow custom implementations
Feature request:
The property "${random.int}" does not help, because it returns any integer (that can be also negative,-2147483648 to 2147483647). I would like to extend the expression, allowing upper and lower bounds to the random, like used in spring. There might be situations, where you need some random values in a given range.

Background:
As an example, where this new "${random.int}" can be used, is for example a cron expression, like `0 ${random.int[0,20]} 2 * * *`, that allows to run the cron at a random hour between 0h and 22h. If you want to have multiple instances of your micronaut microservice, it makes sense, that the cron for example does not run on every instance at the same moment. With the random, you could better achieve, that the processing does not run at same time.

The changes must be done here: https://github.com/micronaut-projects/micronaut-core/blob/2.4.x/inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java

### Ideas

1. allow giving/configuring the range of randoms in properties, like `${random.int[0,20]}`
2. I would also like the approach, if you can add own "property resolver" here, that would allow custom implementations as well. But for now, all related methods are private and not reuseable for a more common approach.

### Expected Behaviour

the property `${random.int[1,7]}` would return a value between 1 and 7 (would also work with long, etc.)

Happy to hear, what you might think about the request.

### PR DESCRIPTION
**Pull request by utsav0209**
Solves issue #5291 

The range can be defined on  `int`, `integer`, `long` and `float`

There are two ways to specify range now as follows: 
1. `random.integer(number)`
2. `random.integer[lower_bound, upper_bound]`

### CODE REVIEW

**Comment by graemerocher** [comment]
Nice work, could you target 2.5.x since this is a new feature and it will be out soon?

**Comment by utsav0209** [comment]
@graemerocher changed the target branch.

**Comment by utsav0209** [comment]
@graemerocher added docs for the same

File: src/main/docs/guide/config/propertySource.adoc [hunk 1]
```diff
@@ -160,6 +160,20 @@ micronaut:
...
+- `[min,max]` where min being inclusive and max being exclusive values.
+
+[source,yaml]
+----
+instance:
+  id: ${random.int[5,10]}
+  count: ${random.int(5)}
+----
+
+NOTE: The range could very from negative to positive as well.
```
Comment (graemerocher) [inline src/main/docs/guide/config/propertySource.adoc]: Should be `vary` instead of `very`

**Review by graemerocher — CHANGES_REQUESTED** [review]

**Review by graemerocher — APPROVED** [review]

File: inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java [hunk 2]
```diff
@@ -885,6 +892,36 @@ private void fill(List list, Integer toIndex, Object value) {
...
+    private float getNextFloatInRange(String range) {
+        String[] tokens = range.split(",");
+        float lowerBound = Float.parseFloat(tokens[0]);
+        if (tokens.length == 1) {
+            return (float) (Math.random() * (lowerBound));
+        }
+        float upperBound = Float.parseFloat(tokens[1]);
+        return lowerBound + (float) (Math.random() * (upperBound - lowerBound));
+    }
+
```
Comment (jameskleeh) [inline inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java]: I think the only thing missing here is better error handling. If the user doesn't use the feature correctly they would get `NumberFormatException`, etc... however that wouldn't be a very helpful error message. This logic should catch any exceptions thrown by `parse` methods and report a more helpful message to indicate to the user what they did wrong.

Comment (utsav0209) [reply inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java]: @jameskleeh is the last commit I pushed is something you want?

Comment (jameskleeh) [reply inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java]: @utsav0209 I would instead throw a ValueException with the original passed in as the cause

Comment (utsav0209) [reply inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java]: @jameskleeh  Okay understood, will make the change and should I keep the Log traces?

Comment (jameskleeh) [reply inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java]: The exception will be logged so no need for additional logs

File: inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java [hunk 3]
```diff
@@ -885,6 +893,48 @@ private void fill(List list, Integer toIndex, Object value) {
...
+        try {
+            String[] tokens = range.split(",");
+            int lowerBound = Integer.parseInt(tokens[0]);
+            if (tokens.length == 1) {
+                return lowerBound >= 0 ? 1 : -1  * (random.nextInt(Math.abs(lowerBound)));
+            }
+            int upperBound = Integer.parseInt(tokens[1]);
+            return lowerBound + (int) (Math.random() * (upperBound - lowerBound));
+        } catch (NumberFormatException ex) {
+            throw new ValueException("Invalid range: `" + range + "` found for type Integer while parsing property: " + property);
```
Comment (jameskleeh) [inline inject/src/main/java/io/micronaut/context/env/PropertySourcePropertyResolver.java]: This is missing the cause

**Comment by graemerocher** [comment]
Thanks for the contribution!

**Pull request merged**